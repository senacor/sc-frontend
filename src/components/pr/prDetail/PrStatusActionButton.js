import React from 'react';
import Button from '@material-ui/core/Button';
import { withStyles } from '@material-ui/core/styles/index';

const styles = theme => ({
  buttonDesktop: {
    position: 'relative',
    backgroundColor: theme.palette.primary['400'],
    color: '#FFF',
    marginBottom: '2%'
  },
  sizeSmall: {}
});

function PrStatusActionButton(props) {
  let { classes, label } = props;
  return (
    <Button
      classes={{ sizeSmall: classes.sizeSmall }}
      size="small"
      className={classes.buttonDesktop}
      onClick={props.releaseButtonClick}
    >
      {label}
    </Button>
  );
}

export default withStyles(styles)(PrStatusActionButton);
