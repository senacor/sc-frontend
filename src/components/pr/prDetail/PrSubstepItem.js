import React from 'react';
import ListItem from '@material-ui/core/ListItem/ListItem';
import ListItemText from '@material-ui/core/ListItemText/ListItemText';
import { withStyles } from '@material-ui/core';

const styles = theme => {
  return {
    listItemRoot: {
      paddingLeft: '0px',
      paddingRight: '0px',
      paddingTop: 0.5 * theme.spacing.unit,
      paddingBottom: 0.5 * theme.spacing.unit
    },
    listItemPrimary: {
      fontSize: '0.875rem'
    }
  };
};

const PrSubstepItem = props => {
  const { classes, substep } = props;
  return (
    <ListItem
      classes={{
        root: classes.listItemRoot
      }}
    >
      <ListItemText
        classes={{
          primary: classes.listItemPrimary
        }}
        primary={substep.label}
        secondary={
          substep.isCompleted
            ? substep.rendering.complete
            : substep.isCurrentUserActionPerformer
            ? substep.rendering.incompleteForActionPerformer
            : substep.rendering.incompleteForNonActionPerformer
        }
      />
    </ListItem>
  );
};

export default withStyles(styles)(PrSubstepItem);
