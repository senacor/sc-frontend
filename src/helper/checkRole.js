import ROLES from './roles';

export const hasRole = (role, userroles) => {
  return userroles.includes(role);
};

export const isEmployee = (userroles = []) => {
  return userroles.includes('PR_Mitarbeiter');
};

export const isSupervisor = userroles => {
  return userroles.includes(ROLES.PR_CST_LEITER);
};
