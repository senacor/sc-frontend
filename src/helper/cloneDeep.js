const cloneDeep = require('lodash/cloneDeep');

export default cloneDeep;
