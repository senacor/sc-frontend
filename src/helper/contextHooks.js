import { useContext } from 'react';
import { ErrorContext, InfoContext, UserinfoContext } from '../components/App';
import ROLES from './roles';

export const useInfoContext = () => {
  const infoContext = useContext(InfoContext.context);
  return {
    context: infoContext,
    msg: msg => {
      infoContext.setValue({
        hasInfos: true,
        messageId: msg
      });
    },
    hide: () => {
      infoContext.setValue({
        hasInfos: false,
        messageId: ''
      });
    }
  };
};

export const useUserinfoContext = () => {
  const userInfoContext = useContext(UserinfoContext.context);
  const { userroles, userinfo } = userInfoContext.value;
  return {
    context: userInfoContext,
    isSupervisorInPr: pr => {
      return userinfo.userId === pr.supervisor.id;
    },
    isReviewerInPr: pr => {
      return userinfo.userId === pr.reviewer.id;
    },
    isOwnerInPr: pr => {
      return pr.employee.id === userinfo.userId;
    },
    hasRoleAdmin: () => {
      return userroles[0] === 'ADMIN';
    },
    hasRoleHr: () => {
      return userroles.includes(ROLES.PERSONAL_DEV);
    },
    hasRoleEmployee: () => {
      return userroles.includes(ROLES.DEVELOPER);
    },
    hasRoleSupervisor: () => {
      return userroles.includes(ROLES.SUPERVISOR);
    }
  };
};

export const useErrorContext = () => {
  const errorContext = useContext(ErrorContext.context);
  return {
    context: errorContext,
    showGeneral: () => {
      errorContext.setValue({
        hasErrors: true,
        messageId: 'message.error'
      });
    },
    hide: () => {
      errorContext.setValue({ hasErrors: false, messageId: '' });
    },
    hasError: error => {
      return errorContext.value.errors && errorContext.value.errors[error];
    },
    show: (msgId, errors) => {
      errorContext.setValue(
        errors
          ? { hasErrors: true, messageId: msgId, errors: errors }
          : { hasErrors: true, messageId: msgId }
      );
    }
  };
};
