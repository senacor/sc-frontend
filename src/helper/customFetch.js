import store from '../store';
import { LOGIN_UNAUTHORIZED } from './dispatchTypes';
/**
 * supported usage:
 *   fetch('URL', configObj);
 *
 *   where configObj is something like:
 *   {
 *     method,
 *     headers,
 *     body
 *   }
 */
export default function customFetch(url, config = {}) {
  let access_token = localStorage.getItem('access_token');

  let authenticationConfig = Object.assign({}, config);
  authenticationConfig.headers = Object.assign({}, config.headers, {
    Authorization: `Bearer ${access_token}`,
    'Content-Type': 'application/json'
  });
  return fetch(url, authenticationConfig).then(response => {
    if (response.status === 401) {
      store.dispatch({
        type: LOGIN_UNAUTHORIZED,
        payload: response.status
      });
    }

    return response;
  });
}

export function fileFetch(url, config = {}) {
  let access_token = localStorage.getItem('access_token');

  let authenticationConfig = Object.assign({}, config);
  authenticationConfig.headers = Object.assign({}, config.headers, {
    Authorization: `Bearer ${access_token}`
  });
  return fetch(url, authenticationConfig).then(response => {
    if (response.status === 401) {
      store.dispatch({
        type: LOGIN_UNAUTHORIZED,
        payload: response.status
      });
    }

    return response;
  });
}
