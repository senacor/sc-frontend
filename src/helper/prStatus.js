export const prStatusEnum = {
  RELEASED_SHEET_REVIEWER: 'FILLED_SHEET_REVIEWER',
  RELEASED_SHEET_EMPLOYEE: 'FILLED_SHEET_EMPLOYEE',
  REQUESTED_DATE: 'REQUESTED_DATE',
  FIXED_DATE: 'ALL_DATES_ACCEPTED',
  FINALIZED_REVIEWER: 'MODIFICATIONS_ACCEPTED_REVIEWER',
  FINALIZED_EMPLOYEE: 'MODIFICATIONS_ACCEPTED_EMPLOYEE',
  ARCHIVED_HR: 'PR_COMPLETED'
};
