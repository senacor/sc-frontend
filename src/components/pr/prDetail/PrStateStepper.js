import React from 'react';
import StepConnector from '@material-ui/core/StepConnector/StepConnector';
import Step from '@material-ui/core/Step/Step';
import StepLabel from '@material-ui/core/StepLabel/StepLabel';
import Typography from '@material-ui/core/Typography/Typography';
import Stepper from '@material-ui/core/Stepper/Stepper';
import { withStyles } from '@material-ui/core';
import List from '@material-ui/core/List';
import PrSubstepItem from './PrSubstepItem';

const styles = theme => {
  return {
    stepper: {
      backgroundColor: 'inherit',
      padding: '0',
      width: '100%'
    },
    mainStepLabel: {
      color: theme.palette.primary['400']
    },
    stepLabelRoot: {
      display: 'flex',
      alignItems: 'flex-start'
    },
    stepperRoot: {
      display: 'flex',
      alignItems: 'flex-start'
    },
    stepConnectorRoot: {
      paddingTop: '10px'
    }
  };
};

const PrStateStepper = ({ classes, stepStructure, activeStep }) => {
  const getExtraStepContent = substeps => {
    return Object.values(substeps).map((substep, index) => {
      return <PrSubstepItem key={`SubStepGrid_${index}`} substep={substep} />;
    });
  };

  return (
    <Stepper
      orientation="horizontal"
      className={classes.stepper}
      activeStep={activeStep}
      classes={{ root: classes.stepperRoot }}
      connector={
        <StepConnector classes={{ root: classes.stepConnectorRoot }} />
      }
    >
      {stepStructure.map((mainStep, index) => {
        return (
          <Step key={mainStep.mainStepLabel}>
            <StepLabel classes={{ root: classes.stepLabelRoot }}>
              <Typography
                variant="subtitle1"
                gutterBottom
                className={classes.mainStepLabel}
              >
                {mainStep.mainStepLabel}
              </Typography>
              {index <= activeStep && (
                <List>{getExtraStepContent(mainStep.substeps)}</List>
              )}
            </StepLabel>
          </Step>
        );
      })}
    </Stepper>
  );
};

export default withStyles(styles)(PrStateStepper);
