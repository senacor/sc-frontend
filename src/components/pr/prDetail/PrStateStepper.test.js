import React from 'react';
import PrStateStepper from './PrStateStepper';
import { createShallow } from '@material-ui/core/test-utils';
import { prStatusEnum } from '../../../helper/prStatus';
import PrStatusActionButton from './PrStatusActionButton';
import PrSubstepItem from './PrSubstepItem';

describe('PerformanceReviewDetail Component', () => {
  let shallow = createShallow({ dive: true });

  const releaseButtonClickMock = jest.fn();
  const activeStep = 1;
  const stepStructure = [
    {
      mainStepLabel: 'Vorbereitung',
      substeps: {
        [prStatusEnum.RELEASED_SHEET_EMPLOYEE]: {
          isCompleted: true,
          isCurrentUserActionPerformer: false,
          label: 'Mitarbeiter: ',
          rendering: {
            complete: 'Abgeschlossen',
            incompleteForNonActionPerformer: 'Nicht abgeschlossen',
            incompleteForActionPerformer: (
              <PrStatusActionButton
                label={'Freigabe'}
                releaseButtonClick={releaseButtonClickMock(
                  prStatusEnum.RELEASED_SHEET_EMPLOYEE
                )}
              />
            )
          }
        },
        [prStatusEnum.RELEASED_SHEET_REVIEWER]: {
          isCompleted: false,
          isCurrentUserActionPerformer: true,
          label: 'Beurteiler: ',
          rendering: {
            complete: 'Abgeschlossen',
            incompleteForNonActionPerformer: 'Nicht abgeschlossen',
            incompleteForActionPerformer: (
              <PrStatusActionButton
                label={'Freigabe'}
                releaseButtonClick={releaseButtonClickMock(
                  prStatusEnum.RELEASED_SHEET_REVIEWER
                )}
              />
            )
          }
        },
        [prStatusEnum.FIXED_DATE]: {
          isCompleted: false,
          isCurrentUserActionPerformer: true,
          label: 'Terminfindung',
          rendering: {
            complete: 'Alle Teilnehmer haben zugesagt',
            incompleteForNonActionPerformer: 'Nicht abgeschlossen',
            incompleteForActionPerformer: 'Nicht abgeschlossen'
          }
        }
      }
    },
    {
      mainStepLabel: 'Gespr√§ch',
      substeps: {
        [prStatusEnum.FINALIZED_REVIEWER]: {
          isCompleted: false,
          isCurrentUserActionPerformer: true,
          label: 'Beurteiler: ',
          rendering: {
            complete: <div>Abgeschlossen</div>,
            incompleteForNonActionPerformer: 'Nicht abgeschlossen',
            incompleteForActionPerformer: (
              <PrStatusActionButton
                label={'Freigabe'}
                releaseButtonClick={releaseButtonClickMock(
                  prStatusEnum.FINALIZED_REVIEWER
                )}
              />
            )
          }
        }
      }
    },
    {
      mainStepLabel: 'Abschluss',
      substeps: {
        [prStatusEnum.FINALIZED_EMPLOYEE]: {
          isCompleted: false,
          isCurrentUserActionPerformer: false,
          label: 'Mitarbeiter:',
          rendering: {
            complete: 'Abgeschlossen',
            incompleteForNonActionPerformer: 'Nicht abgeschlossen',
            incompleteForActionPerformer: (
              <PrStatusActionButton
                label={'Freigabe'}
                releaseButtonClick={releaseButtonClickMock(
                  prStatusEnum.FINALIZED_EMPLOYEE
                )}
              />
            )
          }
        }
      }
    },
    {
      mainStepLabel: 'Archivieren',
      substeps: {
        [prStatusEnum.ARCHIVED_HR]: {
          isCompleted: false,
          label: 'HR: ',
          rendering: {
            complete: 'Archiviert',
            incompleteForNonActionPerformer: 'Nicht abgeschlossen',
            incompleteForActionPerformer: (
              <PrStatusActionButton
                label={'Freigabe'}
                releaseButtonClick={releaseButtonClickMock(
                  prStatusEnum.ARCHIVED_HR
                )}
              />
            )
          }
        }
      }
    }
  ];

  it('should match snapshot', () => {
    const component = shallow(
      <PrStateStepper stepStructure={stepStructure} activeStep={activeStep} />
    );
    expect(component).toMatchSnapshot();
  });

  it('should call getExtraStepContent which returns the correct number of substeps', () => {
    const component = shallow(
      <PrStateStepper stepStructure={stepStructure} activeStep={activeStep} />
    );

    const result = component
      .instance()
      .getExtraStepContent(stepStructure[0].substeps);

    expect(result).toHaveLength(3);
  });

  it('renders correctly', () => {
    const component = shallow(
      <PrStateStepper stepStructure={stepStructure} activeStep={activeStep} />
    );

    expect(component.props().activeStep).toBe(1);
    expect(component.find('WithStyles(Stepper)')).toHaveLength(1);
    expect(component.find('WithStyles(Step)')).toHaveLength(4);
    expect(component.find('WithStyles(StepLabel)')).toHaveLength(4);
    expect(component.find(PrSubstepItem)).toHaveLength(4);
  });

  it('renders the correct amount of substeps', () => {
    const component = shallow(
      <PrStateStepper stepStructure={stepStructure} activeStep={0} />
    );
    expect(component.find(PrSubstepItem)).toHaveLength(3);
  });
});
