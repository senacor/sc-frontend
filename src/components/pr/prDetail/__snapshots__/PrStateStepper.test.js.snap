// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PerformanceReviewDetail Component should match snapshot 1`] = `
<WithStyles(Stepper)
  activeStep={1}
  className="PrStateStepper-stepper-1"
  classes={
    Object {
      "root": "PrStateStepper-stepperRoot-4",
    }
  }
  connector={
    <WithStyles(StepConnector)
      classes={
        Object {
          "root": "PrStateStepper-stepConnectorRoot-5",
        }
      }
    />
  }
  orientation="horizontal"
>
  <WithStyles(Step)
    key="Vorbereitung"
  >
    <WithStyles(StepLabel)
      classes={
        Object {
          "root": "PrStateStepper-stepLabelRoot-3",
        }
      }
    >
      <WithStyles(Typography)
        className="PrStateStepper-mainStepLabel-2"
        gutterBottom={true}
        variant="subheading"
      >
        Vorbereitung
      </WithStyles(Typography)>
      <WithStyles(List)>
        <WithStyles(PrSubstepItem)
          key="SubStepGrid_0"
          substep={
            Object {
              "isCompleted": true,
              "isCurrentUserActionPerformer": false,
              "label": "Mitarbeiter: ",
              "rendering": Object {
                "complete": "Abgeschlossen",
                "incompleteForActionPerformer": <WithStyles(PrStatusActionButton)
                  label="Freigabe"
                  releaseButtonClick={undefined}
            />,
                "incompleteForNonActionPerformer": "Nicht abgeschlossen",
              },
            }
          }
        />
        <WithStyles(PrSubstepItem)
          key="SubStepGrid_1"
          substep={
            Object {
              "isCompleted": false,
              "isCurrentUserActionPerformer": true,
              "label": "Beurteiler: ",
              "rendering": Object {
                "complete": "Abgeschlossen",
                "incompleteForActionPerformer": <WithStyles(PrStatusActionButton)
                  label="Freigabe"
                  releaseButtonClick={undefined}
            />,
                "incompleteForNonActionPerformer": "Nicht abgeschlossen",
              },
            }
          }
        />
        <WithStyles(PrSubstepItem)
          key="SubStepGrid_2"
          substep={
            Object {
              "isCompleted": false,
              "isCurrentUserActionPerformer": true,
              "label": "Terminfindung",
              "rendering": Object {
                "complete": "Alle Teilnehmer haben zugesagt",
                "incompleteForActionPerformer": "Nicht abgeschlossen",
                "incompleteForNonActionPerformer": "Nicht abgeschlossen",
              },
            }
          }
        />
      </WithStyles(List)>
    </WithStyles(StepLabel)>
  </WithStyles(Step)>
  <WithStyles(Step)
    key="Gespräch"
  >
    <WithStyles(StepLabel)
      classes={
        Object {
          "root": "PrStateStepper-stepLabelRoot-3",
        }
      }
    >
      <WithStyles(Typography)
        className="PrStateStepper-mainStepLabel-2"
        gutterBottom={true}
        variant="subheading"
      >
        Gespräch
      </WithStyles(Typography)>
      <WithStyles(List)>
        <WithStyles(PrSubstepItem)
          key="SubStepGrid_0"
          substep={
            Object {
              "isCompleted": false,
              "isCurrentUserActionPerformer": true,
              "label": "Beurteiler: ",
              "rendering": Object {
                "complete": <div>
                  Abgeschlossen
            </div>,
                "incompleteForActionPerformer": <WithStyles(PrStatusActionButton)
                  label="Freigabe"
                  releaseButtonClick={undefined}
            />,
                "incompleteForNonActionPerformer": "Nicht abgeschlossen",
              },
            }
          }
        />
      </WithStyles(List)>
    </WithStyles(StepLabel)>
  </WithStyles(Step)>
  <WithStyles(Step)
    key="Abschluss"
  >
    <WithStyles(StepLabel)
      classes={
        Object {
          "root": "PrStateStepper-stepLabelRoot-3",
        }
      }
    >
      <WithStyles(Typography)
        className="PrStateStepper-mainStepLabel-2"
        gutterBottom={true}
        variant="subheading"
      >
        Abschluss
      </WithStyles(Typography)>
    </WithStyles(StepLabel)>
  </WithStyles(Step)>
  <WithStyles(Step)
    key="Archivieren"
  >
    <WithStyles(StepLabel)
      classes={
        Object {
          "root": "PrStateStepper-stepLabelRoot-3",
        }
      }
    >
      <WithStyles(Typography)
        className="PrStateStepper-mainStepLabel-2"
        gutterBottom={true}
        variant="subheading"
      >
        Archivieren
      </WithStyles(Typography)>
    </WithStyles(StepLabel)>
  </WithStyles(Step)>
</WithStyles(Stepper)>
`;
